### PR TITLE
Story Locking: Fix permission error 

### LINKED ISSUE #8588: Story Locking: Permission error
## Bug Description

I just encountered an issue with the story locking where an author user was not able to get the lock information for a story due to lack of permissions.

The reason for that is that we're getting the `wp:lockuser` information via `_embed` from the users controller. And since the `context` is `edit`, this requires `list_users` permission in WordPress, which only admins have.

Because of this, we don't actually get a user object in `wp:lockuser` but a 403 error due to lack of permissions.

I think the easiest way to solve this would be to not use `_embed` for getting the lock user, but instead make the user ID, name and avatar part of the lock response itself.

## Expected Behaviour

Authors should be able to see lock information from others, not risking issues because of `list_users` checks.

## Steps to Reproduce

1. Edit a story as an administrator
2. In a different browser or an incognito window, log in as an author user
3. Go to the dashboard
4. See that the dashboard does not load due to a JS error
5. Check preloaded API response to see that there's a 403 error.

## Additional Context

- Plugin Version: 1.10-alpha
- WordPress Version: 5.8
- Operating System: macOS
- Browser: Chrome

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

### PR DESCRIPTION
**Pull request by spacedmonkey**
## Summary

Author / contributors could not embed lock users because of a permission error. 

#### Before
Screenshot 2021-08-05 at 12 43 01

#### After
Screenshot 2021-08-05 at 12 42 49

### QA

- [ ] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:
Ensure that post lock experiment is enabled. 

1. Edit a story as an administrator
2. In a different browser or an incognito window, log in as an author user
3. Go to the dashboard

### UAT

- [ ] UAT should use the same steps as above.

This PR can be tested by following these steps:

1.

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This PR contains automated tests (unit, integration, and/or e2e) to verify the code works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #8588

### CODE REVIEW

File: includes/REST_API/Lock_Controller.php [hunk 1]
```diff
@@ -202,7 +202,12 @@ protected function get_lock( $post_id ) {
 	 * @return true|WP_Error True if the request has read access for the item, WP_Error object otherwise.
 	 */
 	public function get_item_permissions_check( $request ) {
-		return $this->parent_controller->update_item_permissions_check( $request );
+		$check = $this->parent_controller->get_item_permissions_check( $request );
```
Comment (spacedmonkey) [inline includes/REST_API/Lock_Controller.php:205]: Check to see if user can access the current item.

File: includes/REST_API/Lock_Controller.php [hunk 2]
```diff
@@ -202,7 +202,12 @@ protected function get_lock( $post_id ) {
...
+		if ( is_wp_error( $check ) ) {
+			return $check;
+		}
+
+		return $this->get_post_type_cap( $this->post_type, 'edit_posts' );
```
Comment (spacedmonkey) [inline includes/REST_API/Lock_Controller.php:210]: Fallback to see if the user can access the post type in general.

File: includes/REST_API/Stories_Controller.php [hunk 3]
```diff
@@ -464,6 +465,11 @@ protected function prepare_links( $post ): array {
 		$links = parent::prepare_links( $post );
 		add_post_type_support( Story_Post_Type::POST_TYPE_SLUG, 'revisions' );
 
+		// Ensure that no data about story locks, leaks publically. 
+		if ( ! $this->get_post_type_cap( $this->post_type, 'edit_posts' ) ) {
+			return $links;
+		}
```
Comment (spacedmonkey) [inline includes/REST_API/Stories_Controller.php]: Do not embed lock / lock user if you can't edit the post type. This should always have been here.

File: tests/phpunit/tests/REST_API/Lock_Controller.php [hunk 4]
```diff
@@ -158,7 +158,7 @@ public function test_get_item_no_perm() {
 		);
 		$request  = new WP_REST_Request( \WP_REST_Server::READABLE, '/web-stories/v1/web-story/' . $story . '/lock' );
 		$response = rest_get_server()->dispatch( $request );
-		$this->assertErrorResponse( 'rest_cannot_edit', $response, 401 );
+		$this->assertErrorResponse( 'rest_forbidden', $response, 401 );
```
Comment (spacedmonkey) [inline tests/phpunit/tests/REST_API/Lock_Controller.php:161]: This is correct error response, as you are not editing.

File: includes/REST_API/Stories_Controller.php [hunk 5]
```diff
@@ -464,6 +465,11 @@ protected function prepare_links( $post ): array {
 		$links = parent::prepare_links( $post );
 		add_post_type_support( Story_Post_Type::POST_TYPE_SLUG, 'revisions' );
 
+		// Ensure that no data about story locks, leaks publically.
```
Comment (swissspidy) [inline includes/REST_API/Stories_Controller.php]:
```suggestion
		// Ensure that no data about story locks leaks publicly. 
```

File: includes/REST_API/Lock_Controller.php (hunk 2 above)
Comment (swissspidy) [reply includes/REST_API/Lock_Controller.php:210]: This should check for the specific capability for the supplied post ID (`$request['id']`)

**Comment by swissspidy** [comment]
I'm a bit confused because I don't see the crash I reported in #8588 on `main` anymore 🤔 

What I noticed while testing this:

In the "All Stories" list table, the story lock is not shown if the user does not have editing permission for this particular post.

So we should check for `current_user_can( 'edit_post', $post_id )` as well.

**Comment by spacedmonkey** [comment]
> So we should check for current_user_can( 'edit_post', $post_id ) as well.

The whole issue was because the contributor doesn't have access to edit the story but needs to view the lock. This should not be added. A check for if the current user can edit any story should do fine.

**Comment by swissspidy** [comment]
> > So we should check for current_user_can( 'edit_post', $post_id ) as well.
> 
> The whole issue was because the contributor doesn't have access to edit the story but needs to view the lock. This should not be added. A check for if the current user can edit any story should do fine.

I don't think they need to view the lock.

The "All Stories" page purposely hides the lock  if the user does not have access to edit this specific story.

We should do the same.

Otherwise this

1. Creates disparity between the "My Stories" and "All Stories" screens
2. Leaks information about who's currently editing a private story
3. Confuses users because it displays a lock even if they cannot edit the story in question

File: includes/REST_API/Lock_Controller.php (hunk 2 above)
Comment (spacedmonkey) [reply includes/REST_API/Lock_Controller.php:210]: No it shouldn't. That is what was happen in `$this->parent_controller->update_item_permissions_check`. [See](https://github.com/WordPress/wordpress-develop/blob/31a6cd2f781ebe12f2858fd389cd95dccb23cbd5/src/wp-includes/rest-api/endpoints/class-wp-rest-posts-controller.php#L1632-L1640). 

To see a lock, you only need to be able to edit an story, not this story.

**Comment by spacedmonkey** [comment]
> I don't think they need to view the lock.

Then we should just close this PR and issue then?

**Comment by spacedmonkey** [comment]
I believe that https://github.com/google/web-stories-wp/pull/8594 fix the original issue that was reported.

**Comment by swissspidy** [comment]
Closing in favor of #8594 in that case

**Pull request closed without merging**